### PR TITLE
[FLINK-19027][network] Assign exclusive buffers to LocalRecoveredInputChannel.

### PR DESCRIPTION
**Pull request by AHeise**
## What is the purpose of the change

If a local channel does not receive a buffer during recovery, a live lock may occur where the local channel cannot recover because it is waiting for a buffer but upstream operators are fully backpressured and cannot release a buffer.
It's important to treat LocalRecoveredInputChannel like RemoteInputChannels and assign exclusive buffers for recovery to avoid this situation.

## Brief change log

- Assign exclusive buffers to LocalRecoveredInputChannel.
- Ensure SingleInputGateTest does not swallow exceptions during cleanup.
- Revert https://github.com/apache/flink/pull/13532 as the test failure actually revealed the live lock.

## Verifying this change

Adjusted a few unit tests.
Already covered and reveleaed by UnalignedCheckpointITCase.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/LocalRecoveredInputChannel.java [hunk 1]
```diff
@@ -42,8 +42,17 @@
 			TaskEventPublisher taskEventPublisher,
 			int initialBackOff,
 			int maxBackoff,
+			int networkBuffersPerChannel,
 			InputChannelMetrics metrics) {
-		super(inputGate, channelIndex, partitionId, initialBackOff, maxBackoff, metrics.getNumBytesInLocalCounter(), metrics.getNumBuffersInLocalCounter());
+		super(
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/LocalRecoveredInputChannel.java:47]: I'm not sure if I understand this bug and the fix. Why is allocating exclusive buffers for `LocalRecoveredInputChannel` fixing the problem? Isn't it just reducing the window for the live lock to happen? What if downstream tasks are scheduled with a significant delay (exclusive buffers assignment happens after upstream tasks already acquired lot's of buffers).

In other words, Isn't this a semi fix for this bug https://issues.apache.org/jira/browse/FLINK-13203

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateTest.java [hunk 2]
```diff
@@ -217,15 +221,11 @@ public void testConcurrentReadStateAndProcessAndClose() throws Exception {
...
-		} finally {
-			executor.shutdown();
+			executor.invokeAll(Arrays.asList(closeTask, readRecoveredStateTask, processStateTask));
+
 			// wait until the internal channel state recover task finishes
-			executor.awaitTermination(60, TimeUnit.SECONDS);
 			assertEquals(totalBuffers, environment.getNetworkBufferPool().getNumberOfAvailableMemorySegments());
 			assertTrue(inputGate.getCloseFuture().isDone());
-
-			environment.close();
```
[the remark this answers is not in the document]
Comment (AHeise) [reply flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateTest.java:228]: `close` is called by the `Closer`.

`shutdown` + `awaitTermination` is simply the wrong method. `invokeAll` is doing what was intended. Could be an extra commit. However, it should then probably be done on all 10 places that use `submitTasksAndWaitForResults`.

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateTest.java [hunk 3]
```diff
@@ -59,26 +59,27 @@
 import org.apache.flink.runtime.shuffle.ShuffleDescriptor;
```
[the remark this answers is not in the document]
Comment (AHeise) [reply flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateTest.java:59]: I didn't even know that double-tags are a thing. :p

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java [hunk 4]
```diff
@@ -247,7 +248,18 @@ public void setup() throws IOException {
...
+			if (closeFuture.isDone()) {
+				return FutureUtils.completedVoidFuture();
+			}
+			for (InputChannel inputChannel : inputChannels.values()) {
+				if (inputChannel instanceof RemoteRecoveredInputChannel) {
+					((RemoteRecoveredInputChannel) inputChannel).assignExclusiveSegments();
+				}
+			}
+		}
+
```
[the remark this answers is not in the document]
Comment (AHeise) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java]: I'll add. In short, the #number of required buffers is now higher than a few tests (and possibly production setups) assume. Without the lazy initialization, you cannot simulate backpressure in a few scenarios as easily.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/LocalRecoveredInputChannel.java (hunk 1 above)
Comment (AHeise) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/LocalRecoveredInputChannel.java:47]: Hm you are right, it doesn't solve it completely after having read the ticket. However, without a solution for FLINK-13203, there will also not be a real solution here.
On the other hand, it's inherently wrong to treat local and remote channels differently during recovery (they even share the same implementation). So this commit is still fixing the issue in a best effort manner and certainly helps to improve build stability, which is an improvement of its own.

**Comment by AHeise** [comment]
Addressed your comments, @pnowojski , please review again.

**Review by pnowojski — APPROVED** [review]
LGTM

**Pull request merged**